src/commands/synthetics/mobile.ts

```diff
@@ -70,114 +79,129 @@ export const uploadMobileApplications = async (
+    }
+
+    if (appUploadResponse.status !== 'pending') {
+      break
+    }
+
+    await wait(APP_UPLOAD_POLLING_INTERVAL)
+  }
+
+  if (appUploadResponse.status === 'complete' && !appUploadResponse.is_valid) {
+    throw new CriticalError(
+      'INVALID_MOBILE_APP',
+      `Mobile application failed validation for reason: ${appUploadResponse.invalid_app_result?.invalid_message}`
+    )
+  }
+
+  if (appUploadResponse.status === 'user_error') {
+    throw new CriticalError(
+      'INVALID_MOBILE_APP_UPLOAD_PARAMETERS',
+      `Mobile application failed validation for reason: ${appUploadResponse.user_error_result?.user_error_message}`
+    )
+  }
+
+  if (appUploadResponse.status === 'error') {
+    throw new CriticalError('UNKNOWN_MOBILE_APP_UPLOAD_FAILURE', `Unknown mobile application upload error.`)
+  }
+
+  return {appUploadResponse, fileName: multipartPresignedUrlsResponse.file_name}
 }
 
-export const uploadApplication = async (
-  api: APIHelper,
-  applicationPathToUpload: string,
-  testApplicationId: string,
-  uploadedApplicationByPath: {[applicationFilePath: string]: {applicationId: string; fileName: string}[]}
-) => {
-  const fileName = await uploadMobileApplications(api, applicationPathToUpload, testApplicationId)
-  if (!(applicationPathToUpload in uploadedApplicationByPath)) {
-    uploadedApplicationByPath[applicationPathToUpload] = []
+export class AppUploadCache {
+  private cache: {[applicationFilePath: string]: {[applicationId: string]: string | undefined}} = {}
+
+  public setAppCacheKeys(
+    triggerConfigs: TriggerConfig[],
+    testsAndConfigsOverride: (TestNotFound | TestSkipped | TestWithOverride)[]
+  ): void {
+    for (const [index, item] of testsAndConfigsOverride.entries()) {
+      if ('test' in item && item.test.type === 'mobile' && !('errorMessage' in item)) {
+        const appId = item.test.options.mobileApplication!.applicationId
+        const userConfigOverride = triggerConfigs[index].config
+        const appPath = userConfigOverride.mobileApplicationVersionFilePath
+        if (appPath && (!this.cache[appPath] || !this.cache[appPath][appId])) {
+          this.cache[appPath] = {
+            ...(this.cache[appPath] || {}),
+            [appId]: undefined,
+          }
+        }
+      }
+    }
+  }
+
+  public getAppsToUpload(): AppUploadDetails[] {
+    const appsToUpload: AppUploadDetails[] = []
+    for (const appPath of Object.keys(this.cache)) {
+      for (const appId of Object.keys(this.cache[appPath])) {
+        appsToUpload.push({appId, appPath})
+      }
+    }
+
+    return appsToUpload
+  }
+
+  public getFileName(appPath: string, appId: string): string | undefined {
+    return this.cache[appPath][appId]
   }
 
-  uploadedApplicationByPath[applicationPathToUpload].push({
-    applicationId: testApplicationId,
-    fileName,
-  })
+  public setFileName(appPath: string, appId: string, fileName: string): void {
+    this.cache[appPath][appId] = fileName
+  }
```

Comment:
`getFileName` makes sense in the upload context, but it wasn't immediately obvious to me what was happening here. 

What do you think of naming like `(g|s)etUploadedAppFileName`? Little verbose, but I feel like it eases the grok a little -- that these are set when and only when the app has been uploaded successfully